debitpolicy/debitpolicy.go

```diff
@@ -0,0 +1,115 @@
+package debitpolicy
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/dogmatiq/dogma"
+	"github.com/dogmatiq/example/messages/commands"
+	"github.com/dogmatiq/example/messages/events"
+)
+
+// debitpolicy is the aggregate root for the debit policy.
+type debitpolicy struct {
+	// DebitAmount is the current sum of debit amounts for the period, in cents.
+	DebitAmount int64
+}
+
+func (p *debitpolicy) ApplyEvent(m dogma.Message) {
+	switch x := m.(type) {
+	case events.AccountDebitedForWithdrawal:
+		p.DebitAmount += x.Amount
+	case events.AccountDebitedForTransfer:
+		p.DebitAmount += x.Amount
+	}
+}
+
+// Aggregate implements the business logic for a debit policy.
+//
+// It centralizes all debits that are applied to an account over a period of
+// time in order to enforce a limited debit policy.
+type Aggregate struct{}
+
+// New returns a new debit policy instance.
+func (Aggregate) New() dogma.AggregateRoot {
+	return &debitpolicy{}
+}
+
+// Configure configures the behaviour of the engine as it relates to this
+// handlers.
+func (Aggregate) Configure(c dogma.AggregateConfigurer) {
+	c.Name("debitpolicy")
+	c.RouteCommandType(commands.CheckWithdrawalAllowedByDebitPolicy{})
+	c.RouteCommandType(commands.CheckTransferAllowedByDebitPolicy{})
```

Comment:
The prefix "Check" on these command names makes them sound like queries.  I don't have a good suggestion right now, but I think it needs to be expressed as something we DO to the policy, rather than something we ASK of the policy. Perhaps this points to the fact that "policy" isn't the right name either, but I'm not sure. Let's keep discussing it.

Reply:
I get what you mean, but it will always be some kind of "asking" the policy right?

Alternate name ideas: ConfirmWithdrawal, VerifyWithdrawal, ValidateWithdrawal, RequestWithdrawal, ...

All names will need to include something about the "DailyDebitPolicy" to be strict and clear while also not conflicting with other message naming.

Reply:
To me `Check...` explicitly sounds like "request/response", it feels like something a process does, not an aggregate. I wonder if we `Submit[Debit]ForDailyLimitApproval` rather than `Check` it? Just throwing that one out there, it doesn't seem great.

I'll see if I can't find some prior art about this kind of stuff. It was Udi Dahan, iirc, that was the one talking about aggregates (sagas, in NServiceBus) as "policies".

Reply:
Given that we're going to need the ability to "compensate" for failed transactions by decrementing, perhaps the commands we want here are `IncrementDailyDebitCounter` and `DecrementDailyDebitCounter`, or something to that effect. These are definitely *actions* not *questions*.

Reply:
Ok yeah. The `SubmitXxx` sounds better that what i had, but I like the `Increment` and `Decrement` even better. It makes sense now that they should mutate the debit aggregate value and compensate it later. That clears up that issue i had earlier about trying to listen to events from the other aggregate.